### PR TITLE
Adjusts the Emag's effect on the Emergency Shuttle

### LINKED ISSUE #247: Emag Override Speed
Emagging a shuttle shouldn't instantly reduce it to 10 seconds, it should either have a delay OR reduce time by a percentage.

### PR DESCRIPTION
**Pull request by etherware-novice**
# About The Pull Request
Adjusts the emag to continuously multiply the shuttle timer, giving crewmembers more than 10 seconds to get on board once an emag is used

## Why It's Good For The Game
Makes the emag less broken, especially for hijack missions
Closes #247 

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

## Changelog

:cl:
balance: tweaked the emag effect in the escape shuttle
/:cl:

### CODE REVIEW

**Comment by nednaZ** [comment]
Please test that your code runs before setting the pull request to "Open"
Use Draft if you need unit tests.

**Comment by etherware-novice** [comment]
sorry! whats a unit test?

**Comment by nednaZ** [comment]
A unit test is what you see down below, the section saying "All checks have passed"
It confirms that your code can be compiled.

File: code/modules/shuttle/emergency.dm [hunk 1]
```diff
@@ -31,6 +31,11 @@
 	var/hijack_hacking = FALSE
 	var/hijack_announce = TRUE
 
+	// measures the last world time the emag event triggered
+	var/emaglast
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:35]: emag_last

File: code/modules/shuttle/emergency.dm [hunk 2]
```diff
@@ -31,6 +31,11 @@
...
+	// the actual time until emag again
+	var/emagcooldown = 20
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:37]: This can be a define as it doesn't change.
#define EMAG_COOLDOWN 2 SECONDS

Then have a section after the code to undefine it:
#undef EMAG_COOLDOWN

File: code/modules/shuttle/emergency.dm [hunk 3]
```diff
@@ -268,30 +291,10 @@
 	log_game("[key_name(user)] has emagged the emergency shuttle in [COORD(src)] [time] seconds before launch.")
 
 	obj_flags |= EMAGGED
-	SSshuttle.emergency.movement_force = list("KNOCKDOWN" = 60, "THROW" = 20)//YOUR PUNY SEATBELTS can SAVE YOU NOW, MORTAL
-	var/datum/species/S = new
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:272]: Why is this removed?

File: code/modules/shuttle/emergency.dm [hunk 4]
```diff
@@ -268,30 +291,10 @@
 	log_game("[key_name(user)] has emagged the emergency shuttle in [COORD(src)] [time] seconds before launch.")
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:291]: Move this to the the emag section so it doesn't spam admin logs.

File: code/modules/shuttle/emergency.dm [hunk 5]
```diff
@@ -175,8 +180,26 @@
...
+
+	// monkestation addition
+	if(obj_flags & EMAGGED && (!emaglast || emagcooldown + emaglast < world.time) && prob(50))
+		var/cut_time = 2 / rand(1, 5)
+
+		if (cut_time * TIME_LEFT <= 11)
+			SSshuttle.emergency.setTimer(ENGINES_START_TIME)  // to make it align with igniting
+		else
+			SSshuttle.emergency.modTimer(cut_time)
+			if (cut_time > 1)  // providing launch updates if it changes (doesnt roll a 2)
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:192]: Since 2 / 2 would be 0, you can have this check for any value.
```suggestion
			if (cut_time)  // providing launch updates if it changes (doesnt roll a 2)
```

File: code/modules/shuttle/emergency.dm [hunk 6]
```diff
@@ -175,8 +180,26 @@
...
 
+
+	// monkestation addition
+	if(obj_flags & EMAGGED && (!emaglast || emagcooldown + emaglast < world.time) && prob(50))
+		var/cut_time = 2 / rand(1, 5)
+
+		if (cut_time * TIME_LEFT <= 11)
+			SSshuttle.emergency.setTimer(ENGINES_START_TIME)  // to make it align with igniting
+		else
+			SSshuttle.emergency.modTimer(cut_time)
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:191]: Move this to the if check, as you don't want it triggering on a 0.

File: code/modules/shuttle/emergency.dm [hunk 7]
```diff
@@ -175,8 +180,26 @@
 	if(ENGINES_STARTED || (!IS_DOCKED))
 		return .
 
+
+	// monkestation addition
+	if(obj_flags & EMAGGED && (!emaglast || emagcooldown + emaglast < world.time) && prob(50))
+		var/cut_time = 2 / rand(1, 5)
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:186]: As it stands, this leaves the options of: 1(unchanged), 0(unchanged), 0.66(very slight), 0.5(slight) and 0.4(slight)
I'd suggest simply making a pick proc instead of this random setup.

**Review by nednaZ — CHANGES_REQUESTED** [review]
Some issues

File: code/modules/shuttle/emergency.dm (hunk 5 above)
Comment (etherware-novice) [reply code/modules/shuttle/emergency.dm:192]: i thought 2/2 was 1?

File: code/modules/shuttle/emergency.dm (hunk 7 above)
Comment (etherware-novice) [reply code/modules/shuttle/emergency.dm:186]: how would i go about doing that?

File: code/modules/shuttle/emergency.dm (hunk 6 above)
Comment (etherware-novice) [reply code/modules/shuttle/emergency.dm:191]: what do you mean?

File: code/modules/shuttle/emergency.dm [hunk 8]
```diff
@@ -161,6 +166,21 @@
 /obj/machinery/computer/emergency_shuttle/proc/clear_recent_action(mob/user)
 	acted_recently -= user
 
+/obj/machinery/computer/emergency_shuttle/proc/pick_emag_amount()
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:169]: This doesn't need to be a separate proc, the code here can be put into the main proc it's used in.

File: code/modules/shuttle/emergency.dm [hunk 9]
```diff
@@ -10,6 +10,8 @@
 #define STAGE_3 3
 #define STAGE_4 4
 #define HIJACKED 5
+// the actual time until emag again
+#define EMAGCOOLDOWN 20
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:14]: EMAG_COOLDOWN

File: code/modules/shuttle/emergency.dm [hunk 10]
```diff
@@ -161,6 +166,21 @@
 /obj/machinery/computer/emergency_shuttle/proc/clear_recent_action(mob/user)
 	acted_recently -= user
 
+/obj/machinery/computer/emergency_shuttle/proc/pick_emag_amount()
+	switch(rand(1, 5))
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:170]: The numbers below should be defines and undefines for readability.
Instead of "2", it can be #define DOUBLE_TIME 2
Things along those lines.

**Review by nednaZ — CHANGES_REQUESTED** [review]
Fix the catgloves changes too

**Review by nednaZ — CHANGES_REQUESTED** [review]
This currently doesn't even work.
Emagging the shuttle goes straight to the 10 second launch.



ADMIN LOG: Emagged emergency shuttle will launch in 1186 seconds.
SYSTEM ERROR:
The emergency shuttle will launch in 10 seconds

**Comment by etherware-novice** [comment]
..what

File: code/modules/shuttle/emergency.dm [hunk 11]
```diff
@@ -10,6 +10,15 @@
...
 #define HIJACKED 5
+// the actual time until emag again
+#define EMAG_COOLDOWN 15
+
+// various possibilities for how much it change
+#define EMAG_DOUBLE 2
+#define EMAG_STAY 1
+#define EMAG_SLIGHT 0.6
+#define EMAG_HEAVY 0.2
+#define EMAG_OHNO 0.1
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:21]: Too low.
That takes a 3 minute shuttle to 18 seconds. Might as well just keep the old one.

I'd suggest .25

File: code/modules/shuttle/emergency.dm [hunk 12]
```diff
@@ -10,6 +10,15 @@
...
 #define STAGE_4 4
 #define HIJACKED 5
+// the actual time until emag again
+#define EMAG_COOLDOWN 15
+
+// various possibilities for how much it change
+#define EMAG_DOUBLE 2
+#define EMAG_STAY 1
+#define EMAG_SLIGHT 0.6
+#define EMAG_HEAVY 0.2
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:20]: Takes a 3 minute shuttle to 36 seconds.
I'd suggest .35

File: code/modules/shuttle/emergency.dm [hunk 13]
```diff
@@ -175,8 +190,29 @@
 	if(ENGINES_STARTED || (!IS_DOCKED))
 		return .
 
+
+	// monkestation addition
+	if(obj_flags & EMAGGED && (!emag_last || EMAG_COOLDOWN + emag_last < world.time) && prob(40))
+		var/cut_time = pick(EMAG_DOUBLE, EMAG_STAY, EMAG_SLIGHT, EMAG_HEAVY, EMAG_OHNO)
+
+		if (cut_time * TIME_LEFT <= 11)
+			SSshuttle.emergency.setTimer(ENGINES_START_TIME)  // to make it align with igniting
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:199]: Needs an alert message here too.
Make this the old ten second launch message.

File: code/modules/shuttle/emergency.dm [hunk 14]
```diff
@@ -10,6 +10,15 @@
 #define STAGE_3 3
 #define STAGE_4 4
 #define HIJACKED 5
+// the actual time until emag again
+#define EMAG_COOLDOWN 15
```
Comment (nednaZ) [inline code/modules/shuttle/emergency.dm:14]:
```suggestion
#define EMAG_COOLDOWN 1.5 SECONDS
```

**Review by nednaZ — CHANGES_REQUESTED** [review]
Almost ready

File: code/modules/shuttle/emergency.dm [hunk 15]
```diff
@@ -259,16 +296,14 @@
...
 		return
 
-	var/time = TIME_LEFT
-	message_admins("[ADMIN_LOOKUPFLW(user.client)] has emagged the emergency shuttle, [time] seconds before launch.")
-	log_game("[key_name(user)] has emagged the emergency shuttle in [COORD(src)] [time] seconds before launch.")
+
 
 	obj_flags |= EMAGGED
-	SSshuttle.emergency.movement_force = list("KNOCKDOWN" = 60, "THROW" = 20)//YOUR PUNY SEATBELTS can SAVE YOU NOW, MORTAL
+	SSshuttle.emergency.movement_force = list("KNOCKDOWN" = 60, "THROW" = 20) //YOUR PUNY SEATBELTS can SAVE YOU NOW, MORTAL
```
Comment (etherware-novice) [inline code/modules/shuttle/emergency.dm:306]: was getting compile errors with this comment so added whitespace to be certain
